server/src/main/java/com/paypal/selion/proxy/SeLionSauceProxy.java

```diff
@@ -22,72 +22,243 @@
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.logging.Level;
 
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
 import com.paypal.selion.grid.servlets.SauceServlet;
+import com.paypal.selion.utils.SauceConfigReader;
 import com.paypal.selion.utils.SauceLabsRestApi;
 
 import org.apache.commons.lang.StringUtils;
 import org.openqa.grid.common.RegistrationRequest;
 import org.openqa.grid.common.exception.GridConfigurationException;
 import org.openqa.grid.common.exception.GridException;
-import org.openqa.grid.internal.BaseRemoteProxy;
+import org.openqa.grid.common.exception.RemoteNotReachableException;
 import org.openqa.grid.internal.Registry;
 import org.openqa.grid.internal.TestSession;
+import org.openqa.grid.selenium.proxy.DefaultRemoteProxy;
+import org.openqa.selenium.remote.http.HttpMethod;
 
 import com.paypal.selion.logging.SeLionGridLogger;
 
 /**
- * This customized version of {@link BaseRemoteProxy} basically helps redirect all traffic to the SauceLabs cloud.
+ * This customized version of {@link DefaultRemoteProxy} basically helps redirect all traffic to the SauceLabs cloud.
  */
-public class SeLionSauceProxy extends BaseRemoteProxy {
+public class SeLionSauceProxy extends DefaultRemoteProxy {
 
     private static final SeLionGridLogger LOGGER = SeLionGridLogger.getLogger(SeLionSauceProxy.class);
+
+    /**
+     * Sauce labs DesiredCapabilities that {@link SeLionSauceProxy} is concerned with
+     */
+    public static class SauceLabsCapability {
+        public static final String SAUCE_USER_NAME_CAPABILITY = "sauceUserName";
+        public static final String SAUCE_API_KEY_CAPABILITY = "sauceApiKey"; 
+        public static final String USERNAME_CAPABILITY = "username";
+        public static final String ACCESS_KEY_CAPABILITY = "accessKey";
+        public static final String PARENT_TUNNEL_CAPABILITY = "parent-tunnel";
+        public static final String TUNNEL_IDENTIFIER_CAPABILITY = "tunnelIdentifier";
+        public static final String IDLE_TIMEOUT_CAPABILITY = "idleTimeout";
+    }
+
+    /** {@link SauceConfigReader} instruction to use the provided username value also for the parent-tunnel, if omitted */
+    public static final String DEFAULT_PARENT_TUNNEL_TO_USERNAME_TOKEN = "$username";
+    /** Sauce labs cloud default idle timeout is 90 seconds */
+    public static final int DEFAULT_IDLE_TIMEOUT = 90000;
+    /** Overrides value in DefaultRemoteProxy */
+    public static final int DEFAULT_POLLING_INTERVAL = 60000;
+    /** Overrides value in DefaultRemoteProxy */
+    public static final int DEFAULT_DOWN_POLLING_LIMIT = 3;
+
     private final SauceLabsRestApi sauceApi;
+    private final String defaultParentTunnel;
+    private final String defaultTunnelIdentifier;
     private final Lock accessLock = new ReentrantLock();
+    private final SauceConfigReader sauceConfigReader;
+    private Thread pollingThread;
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://www.codacy.com/app/SeLion/paypal-SeLion/file/2414240556/issues/source?bid=2928425&fileBranchId=3233060#l78)